File: 백준/Gold/15681. 트리와 쿼리/트리와 쿼리.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split("\n");

const [N, root, Q] = input[0].split(" ").map(Number);
const edges = input.slice(1, N).map((line) => line.split(" ").map(Number));
const queries = input.slice(N).map((line) => line.split(" ").map(Number));

const adj = Array.from({ length: N + 1 }, () => []);
const size = Array(N + 1).fill(0);

for (const [u, v] of edges) {
  adj[u].push(v);
  adj[v].push(u);
}

function dfs(node, parent) {
  size[node] = 1;

  for (const neighbor of adj[node]) {
    if (neighbor !== parent) {
      dfs(neighbor, node);
      size[node] += size[neighbor];
    }
  }
}

dfs(root, -1);
console.log(queries.map(([u]) => size[u]).join("\n"));

File: 백준/Gold/15681. 트리와 쿼리/트리와 쿼리.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline

def solve():
    n, r, q = map(int, input().split())

    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        u, v = map(int, input().split())
        adj[u].append(v)
        adj[v].append(u)
    
    size = [0] * (n + 1)

    def count_subtree_nodes(cur, parent):
        size[cur] = 1

        for neighbor in adj[cur]:
            if neighbor != parent:
                count_subtree_nodes(neighbor, cur)
                size[cur] += size[neighbor]
    
    count_subtree_nodes(r, -1)

    for _ in range(q):
        u = int(input())
        print(size[u])

if __name__ == "__main__":
    solve()